### ISSUE TITLE
Has anybody implement QUAD method?

### ISSUE THREAD

**Issue by jjprincess** [argman/EAST#167]
I've tried to do QUAD method in this paper,can anybody help me ,thank you very much!

**Comment by Yuanhang8605** [comment]
I have implemented quad mode, predict the offset relative to the positive pixel, use the smooth L1 loss

**Comment by Yuanhang8605** [comment]
@jjprincess, here is my tensorflow code : 
  def _localization_loss(self, pred_geo_map, 
                         gt_geo_map, gt_score_map):
    """Cal the localization loss. 
    """
    # the point order: x1, y1, x2, y2, x3, y3, x4, y4
    # smooth_L1 loss. 
    gt_geo_map = gt_geo_map / float(self._text_scale)     # I set _text_scale = 50 to shrink the gt_geo_map. 
    gt_min_edge = gt_geo_map[:, :, :, -1]
    gt_min_edge = tf.expand_dims(gt_min_edge, 3)
    diff = pred_geo_map - gt_geo_map[:, :, :, :-1]

    abs_diff = tf.abs(diff)
    abs_diff_lt_1 = tf.less(abs_diff, 1)
    L_g = tf.where(abs_diff_lt_1, 0.5 * tf.square(abs_diff), abs_diff - 0.5) 
    L_g = L_g / gt_min_edge
    L_g = tf.reduce_mean(L_g, axis=3)

**Comment by jjprincess** [comment]
@Yuanhang8605 Thanks, I has implemented this with caffe too.^_^

**Comment by xuannianc** [comment]
@jjprincess thanks for the sharing, what does self._text_scale use for and why do you set gt_min_edge as the last element of gt_geo_map?

**Comment by xuannianc** [comment]
@jjprincess I also got very high loss using N<sub>Q<sup>*</sup></sub> mentioned in the paper, can you share your code for my reference?
@Yuanhang8605 Do you shrink by self._text_scale to fix this?

**Comment by uname0x96** [comment]
> @jjprincess, here is my tensorflow code :
> 
> ```python
>   def _localization_loss(self, pred_geo_map, 
>                          gt_geo_map, gt_score_map):
>     """Cal the localization loss. 
>     """
>     # the point order: x1, y1, x2, y2, x3, y3, x4, y4
>     # smooth_L1 loss. 
>     gt_geo_map = gt_geo_map / float(self._text_scale)     # I set _text_scale = 50 to shrink the gt_geo_map. 
>     gt_min_edge = gt_geo_map[:, :, :, -1]
>     gt_min_edge = tf.expand_dims(gt_min_edge, 3)
>     diff = pred_geo_map - gt_geo_map[:, :, :, :-1]
> 
>     abs_diff = tf.abs(diff)
>     abs_diff_lt_1 = tf.less(abs_diff, 1)
>     L_g = tf.where(abs_diff_lt_1, 0.5 * tf.square(abs_diff), abs_diff - 0.5) 
>     L_g = L_g / gt_min_edge
>     L_g = tf.reduce_mean(L_g, axis=3)
> ```

Hi @Yuanhang8605 , can you share your code for quad box? i has implemented but got the high loss, seem it not working

**Comment by Adnan-annan** [comment]
> > @jjprincess, here is my tensorflow code :
> > ```python
> >   def _localization_loss(self, pred_geo_map, 
> >                          gt_geo_map, gt_score_map):
> >     """Cal the localization loss. 
> >     """
> >     # the point order: x1, y1, x2, y2, x3, y3, x4, y4
> >     # smooth_L1 loss. 
> >     gt_geo_map = gt_geo_map / float(self._text_scale)     # I set _text_scale = 50 to shrink the gt_geo_map. 
> >     gt_min_edge = gt_geo_map[:, :, :, -1]
> >     gt_min_edge = tf.expand_dims(gt_min_edge, 3)
> >     diff = pred_geo_map - gt_geo_map[:, :, :, :-1]
> > 
> >     abs_diff = tf.abs(diff)
> >     abs_diff_lt_1 = tf.less(abs_diff, 1)
> >     L_g = tf.where(abs_diff_lt_1, 0.5 * tf.square(abs_diff), abs_diff - 0.5) 
> >     L_g = L_g / gt_min_edge
> >     L_g = tf.reduce_mean(L_g, axis=3)
> > ```
> 
> Hi @Yuanhang8605 , can you share your code for quad box? i has implemented but got the high loss, seem it not working

Hi bro,, i am trying to implement QUAD geometry for this tensorflow EAST code. can you share your working code with me ?

**Comment by Adnan-annan** [comment]
> @Yuanhang8605 Thanks, I has implemented this with caffe too.^_^

Hi bro how did you fix the high loss problem ?

**Comment by Adnan-annan** [comment]
> @jjprincess, here is my tensorflow code :
> 
> ```python
>   def _localization_loss(self, pred_geo_map, 
>                          gt_geo_map, gt_score_map):
>     """Cal the localization loss. 
>     """
>     # the point order: x1, y1, x2, y2, x3, y3, x4, y4
>     # smooth_L1 loss. 
>     gt_geo_map = gt_geo_map / float(self._text_scale)     # I set _text_scale = 50 to shrink the gt_geo_map. 
>     gt_min_edge = gt_geo_map[:, :, :, -1]
>     gt_min_edge = tf.expand_dims(gt_min_edge, 3)
>     diff = pred_geo_map - gt_geo_map[:, :, :, :-1]
> 
>     abs_diff = tf.abs(diff)
>     abs_diff_lt_1 = tf.less(abs_diff, 1)
>     L_g = tf.where(abs_diff_lt_1, 0.5 * tf.square(abs_diff), abs_diff - 0.5) 
>     L_g = L_g / gt_min_edge
>     L_g = tf.reduce_mean(L_g, axis=3)
> ```

Hi bro. could you guide me where in multigpu_train.py i need to adjust this code for QUAD ?

**Comment by Adnan-annan** [comment]
> @Yuanhang8605 Thanks, I has implemented this with caffe too.^_^

How to modify the geometry map generation for QUAD?

**Comment by ankschin** [comment]
Hello everyone,
I have tried implementing above mentioned code as well as my own code also in a different way. But after 1st training point the loss changes to Nan.
After debugging I saw all the weight and last layer output all are Nan. Please help in understanding what could have gone wrong.

**Comment by ankschin** [comment]
> @jjprincess, here is my tensorflow code :
> 
> ```python
>   def _localization_loss(self, pred_geo_map, 
>                          gt_geo_map, gt_score_map):
>     """Cal the localization loss. 
>     """
>     # the point order: x1, y1, x2, y2, x3, y3, x4, y4
>     # smooth_L1 loss. 
>     gt_geo_map = gt_geo_map / float(self._text_scale)     # I set _text_scale = 50 to shrink the gt_geo_map. 
>     gt_min_edge = gt_geo_map[:, :, :, -1]
>     gt_min_edge = tf.expand_dims(gt_min_edge, 3)
>     diff = pred_geo_map - gt_geo_map[:, :, :, :-1]
> 
>     abs_diff = tf.abs(diff)
>     abs_diff_lt_1 = tf.less(abs_diff, 1)
>     L_g = tf.where(abs_diff_lt_1, 0.5 * tf.square(abs_diff), abs_diff - 0.5) 
>     L_g = L_g / gt_min_edge
>     L_g = tf.reduce_mean(L_g, axis=3)
> ```

I think these two lines just gives the last value y4 instead of smallest edge. Please correct me if I am wrong.
  gt_min_edge = gt_geo_map[:, :, :, -1]
  gt_min_edge = tf.expand_dims(gt_min_edge, 3)
